projects/bootloader/hal-drivers.cmake

```diff
@@ -0,0 +1,58 @@
+# List of HAL source files
+set(HAL_DRIVERS_SRC_FILES
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim_ex.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc_ex.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ex.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ramfunc.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_gpio.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma_ex.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr_ex.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_cortex.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_exti.c
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_crc.c
+)
+
+add_library(hal_drivers ${HAL_DRIVERS_SRC_FILES})
+
+# List of compiler defines, prefix with -D compiler option
+target_compile_definitions(hal_drivers PUBLIC
+        #-DUSE_HAL_DRIVER
+        -DSTM32F401xE
+        )
+
+target_include_directories(hal_drivers PUBLIC
+    ${GIT_ROOT_DIR}/projects/bootloader/boards/stm32f401re/Inc
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Inc
+    ${GIT_ROOT_DIR}/third_party/Drivers/STM32F4xx_HAL_Driver/Inc/Legacy
+    ${GIT_ROOT_DIR}/third_party/Drivers/CMSIS/Device/ST/STM32F4xx/Include
+    ${GIT_ROOT_DIR}/third_party/Drivers/CMSIS/Include
+)
+
+target_compile_options(hal_drivers PRIVATE -Wno-unused-parameter -mthumb
```

Comment:
I think this is duplicate code. I am not sure how we should handle the compiler/link options in the cmake, but I feel like we can do it at one place (in the CMakeLists file. Can you please figure it out and clean it up? If it is already in your TODO list for next PR, feel free to skip this comment.

Reply:
This file provides guidelines for complier and linker on how to deal with the HAL library only, thus it is not duplicate. It extends the possibility of compiling and linking this specific library using different options than the rest of the code. Although it is arguable if it makes sense for embedded applications, for non-embedded ones it is very useful to have.